modules/tinymce/src/core/test/ts/browser/util/BeforeInputEventUtils.ts

```diff
@@ -0,0 +1,49 @@
+import { TinyAssertions, TinyContentActions, TinySelections } from '@ephox/wrap-mcagar';
+import { assert } from 'chai';
+
+import Editor from 'tinymce/core/api/Editor';
+import { EditorEvent } from 'tinymce/core/api/util/EventDispatcher';
+import * as InsertNewLine from 'tinymce/core/newline/InsertNewLine';
+import Env from 'tinymce/src/core/main/ts/api/Env';
+import { NormalizedEvent } from 'tinymce/src/core/main/ts/events/EventUtils';
+
+export const testBeforeInputEvent = (performEditAction: (editor: Editor) => void, eventType: string, expectMoreEventsIfSafari: boolean) =>
+  (
+    editor: Editor,
+    setupHtml: string,
+    setupPath: number[],
+    setupOffset: number,
+    expectedHtml: string,
+    cancelBeforeInput: boolean
+  ) => {
+    const inputEvents: string[] = [];
+    const collect = (event: NormalizedEvent<InputEvent, any>) => {
+      inputEvents.push(event.inputType);
+    };
+    const beforeInputCollect = (event: NormalizedEvent<InputEvent, any>) => {
+      collect(event);
+
+      if (cancelBeforeInput) {
+        event.preventDefault();
+      }
+    };
+
+    editor.on('input', collect);
+    editor.on('beforeinput', beforeInputCollect);
+    editor.setContent(setupHtml);
+    TinySelections.setCursor(editor, setupPath, setupOffset);
+    performEditAction(editor);
+    editor.off('beforeinput', beforeInputCollect);
+    editor.off('input', collect);
+
+    TinyAssertions.assertContent(editor, expectedHtml);
+
+    const moreEvents = expectMoreEventsIfSafari && Env.browser.isSafari();
+    const expectedNoCancelEvents = moreEvents ? [ eventType, eventType, eventType ] : [ eventType, eventType ];
```

Comment:
This is not the right approach to this patching over the test so that it passes even when it's sending multiple events is not the right way to solve this. We need to figure out why these extra events are being sent. The patch I sent you should work around that I will double check why you couldn't get it to work since it passed on my local Safari.

Reply:
Ok worked some on this. I made an error in the patch I sent you and only tested the beforeinput tests sorry about that. I corrected that and removed the Safari workarounds and all tests now pass on CI. I put that in a separate branch you can cherry pick the HEAD of that. 

- [ ] Cherry pick from https://github.com/tinymce/tinymce/tree/feature/TINY-8329_2

Reply:
Are we sure this is the solution we want to go for? This issue about double-firing the input event feels like it's a bug in some unrelated code, and should be fixed separately rather than trying to patch the signs of it like this?

Reply:
Yes, this is all about normalizing browser behavior and not firing multiple events. Our custom delete logic should just fire one `beforeInput` and one `input` event. Safari seems to be firing both `beforeInput` and `input` events when doing for example `document.execCommand('Delete')` other browsers seems to only fire the `input` event. So we could just use the same method that we used before for preventing double `input` events from being emitted by the native `execCommand` call by also include the `beforeinput` events to the stop the propagation to other handlers.

So we know why they are emitted and how to prevent that form getting handled by the tinymce event system. So that is what the patch I provided fixes rather than fixing the tests. The tests should only be branched based on browsers if it doesn't really matter if one browser does one thing and another browser does another thing or as a last resort if there really is no workaround for a particular browser. In this case it matters we want to avoid emitting multiple events and we can also work around it.

I also separated the `beforeinput` and `input` events into two buckets as mentioned before we can't put them together into one bucket for assertion since then we don't know what event added that particular input event type. It could be that `beforeinput` fired twice or `input` fired twice or `beforeinput` and `input` fired once all would result in a single array with two items in it. We want to check for the latter that beforeinput fired and input fired once.